src/ints/bios_pci.cpp

```diff
@@ -0,0 +1,265 @@
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "bios.h"
+
+#include "callback.h"
+#include "checks.h"
+#include "dosbox.h"
+#include "inout.h"
+#include "pci_bus.h"
+#include "regs.h"
+#include "support.h"
+
+CHECK_NARROWING();
+
+// Reference:
+// - PCI BIOS Specification, revision 2.1
+
+// TODO (not sure if needed by anything):
+// - interrupt routing is not implemented
+// - special cycles are not implemented
+// - BIOS32 interface is not implemented
+// - 0x000ffe6e entry point is not implemented
+
+constexpr uint16_t max_device_index = 0x100;
+
+enum class ReturnCode : uint8_t {
+	Successful        = 0x00,
+	FuncNotSupported  = 0x81,
+	BadVendorId       = 0x83,
+	DeviceNotFound    = 0x86,
+	BadRegisterNumber = 0x87,
+	SetFailed         = 0x88,
+	BufferTooSmall    = 0x89,
+};
+
+static void warn_unknown_function(const uint8_t function)
+{
+	static bool already_warned[UINT8_MAX + 1];
+	if (!already_warned[function]) {
+		LOG_WARNING("INT1A: Unknown PCI function 0xb1%02x", function);
+		already_warned[function] = true;
+	}
+}
+
+static void warn_no_pci_present()
+{
+	static bool already_warned = false;
+	if (!already_warned) {
+		LOG_WARNING("INT1A: PCI function called despite no PCI present");
+		already_warned = true;
+	}
+}
+
+void INT1AB1_Handler()
+{
+	auto set_return_code = [](const ReturnCode code) {
+		reg_ah = enum_val(code);
+		CALLBACK_SCF(code != ReturnCode::Successful);
+	};
+
+	auto select_read_write_address = []() {
+		const auto address = 0x80000000 |
+	                             static_cast<uint32_t>(reg_bx << 8) |
+	                             static_cast<uint32_t>(reg_di & 0xfc);
+		IO_WriteD(port_num_pci_config_address,
+		          static_cast<uint32_t>(address));
+	};
+
+	if (!PCI_IsInitialized()) {
+		// No PCI subsystem
+
+		if (reg_al != 0x01) {
+			warn_no_pci_present();
+		}
+
+		set_return_code(ReturnCode::FuncNotSupported);
+		return;
+	}
+
+	switch (reg_al) {
+	case 0x01: // PCI BIOS Present
+		reg_bx  = 0x0210;     // version 2.10
+		reg_cx  = 0x0000;     // only one PCI bus
+		reg_edx = 0x20494350; // "PCI "
+		reg_edi = PCI_GetPModeInterface();
+		// AL informs which mechanisms are supported:
+		// bit 0: configuration mechanism #1
+		// bit 1: configuration mechanism #2
+		// (either bit 1 or 2 needs to be set)
+		// bit 4: special cycle generation mechanism #1
+		// bit 5: special cycle generation mechanism #2
+		// (either bit 4 or 5 can be set, must match the supported
+		// configuration mechanism)
+		reg_al = 0x01;
+		set_return_code(ReturnCode::Successful);
+		return;
+	case 0x02: // Find PCI Device
+	{
+		// Check if vendor ID is valid
+		if (reg_dx == UINT16_MAX) {
```

Comment:
Dunno, wouldn't `0xffff` be better here? We're not really checking for the "biggest number", but for a magic value, no?

Reply:
OK, changed.